src/lib/rpc/bestRpcTracker.ts

```diff
@@ -0,0 +1,368 @@
+import { Provider, ethers } from "ethers";
+import {
+  RPC_PROVIDERS,
+  SUPPORTED_CHAIN_IDS,
+  ARBITRUM,
+  AVALANCHE,
+  AVALANCHE_FUJI,
+  FALLBACK_PROVIDERS,
+} from "config/chains";
+import { getRpcProviderKey, SHOW_DEBUG_VALUES_KEY } from "config/localStorage";
+import entries from "lodash/entries";
+import orderBy from "lodash/orderBy";
+import meanBy from "lodash/meanBy";
+import minBy from "lodash/minBy";
+import sample from "lodash/sample";
+import { differenceInMilliseconds } from "date-fns";
+import { getMulticallContract, getDataStoreContract } from "config/contracts";
+import { getContract } from "config/contracts";
+import { HASHED_MARKET_CONFIG_KEYS } from "prebuilt";
+import { getIsFlagEnabled } from "config/ab";
+import { sleep } from "lib/sleep";
+
+const PROBE_INTERVAL = 10 * 1000; // 10 seconds / Frequency of RPC probing
+const PROBE_FAIL_TIMEOUT = 10 * 1000; // 10 seconds / Abort RPC probe if it takes longer
+const PROBE_SLIDING_WINDOW = 60 * 1000; // 60 seconds / Size of the sliding window for RPC average response time calculation
+const STORAGE_EXPIRE_TIMEOUT = 5 * 60 * 1000; // 5 minutes / Time after which provider saved in the localStorage is considered stale
+const DISABLE_UNUSED_TRACKING_TIMEOUT = 2 * 60 * 1000; // 2 minutes / Pause probing if no requests for the best RPC for this time
+
+const BLOCK_FROM_FUTURE_THRESHOLD = 1000; // Omit RPC if block number is higher than average on this value
+const BLOCK_LAGGING_THRESHOLD = 50; // Omit RPC if block number is lower than highest valid on this value
+
+// DataStore field used for probing
+const PROBE_SAMPLE_FIELD = "minCollateralFactor";
+// Markets used for `PROBE_SAMPLE_FIELD` reading
+const PROBE_SAMPLE_MARKET = {
+  [ARBITRUM]: "0x70d95587d40A2caf56bd97485aB3Eec10Bee6336", // ETH/USD
+  [AVALANCHE]: "0xB7e69749E3d2EDd90ea59A4932EFEa2D41E245d7", // ETH/USD
+  [AVALANCHE_FUJI]: "0xbf338a6C595f06B7Cfff2FA8c958d49201466374", // ETH/USD
+};
+
+type RpcProbe = {
+  isSuccess: boolean;
+  responseTime: number;
+  blockNumber: number | null;
+  timestamp: Date;
+};
+
+type ProviderData = {
+  url: string;
+  provider: Provider;
+  probeData: RpcProbe[];
+};
+
+type RpcTrackerState = {
+  [chainId: number]: {
+    lastUsage: Date;
+    currentBestProviderUrl: string;
+    providers: {
+      [providerUrl: string]: ProviderData;
+    };
+  };
+};
+
+let trackingIntervalId: NodeJS.Timeout | null = null;
```

Comment:
you can use `window.setInterval` instead of just `setInterval` that way you would have `number` instead of nodejs type